python/pyproject.toml

```diff
@@ -0,0 +1,47 @@
+# The DOLFINx Python interface must be built without build isolation (PEP517)
+# due to its depedency on system built petsc4py and mpi4py.
+# pip install -r build_requirements.txt
+[build-system]
+requires = ["scikit-build-core[pyproject]", "pybind11>=2.9.1", "petsc4py", "mpi4py"]
```

Comment:
Is this correct? I would say we don't require python basix to build dolfinx python wrappers.

Reply:
Not sure what the question exactly is, but I do not follow why `petsc4py` and `mpi4py` are both here and in `dependencies`.

We should probably be aware of https://github.com/pypa/pip/issues/9314
which then led to the introduction of `--check-build-dependencies` in https://github.com/pypa/pip/issues/9794 and https://github.com/pypa/pip/pull/10886: due to `--no-build-isolation` the addition of  `petsc4py` and `mpi4py` to `requires` will only matter for checking build dependencies, not installing them.

Reply:
petsc4py and mpi4py are build time and runtime dependencies of the python interface, so are listed in both places. pybind11 is only a build time dependency.

We can leave petsc4py and mpi4py out but that gives a much more cryptic error message from cmake.

Reply:
As a downstream maintainer, I'd appreciate it if build dependencies are explicit. In conda, we don't use pip isolated builds (because conda is responsible for the env), but build dependencies and runtime dependencies are separate environments, and it can be beneficial to avoid putting unused runtime dependencies in the build environment (saves a small amount of time, but can also save unnecessarily strict dependency pinning at runtime).

Reply:
From @jhale's answer I understood that `mpi4py` and `petsc4py` are explicitly put in the build dependencies because `cmake` is looking for them at build time, right? They are both obviously also runtime dependencies, since the library imports them.

Reply:
`mpi4py` and `petsc4py` *are* build dependencies. Maybe not immediately obvious, but we need the `mpi4py` and `petsc4py` header files for the C++ <-> Python casters.

Reply:
I have improved the comment to mention buildtime and runtime.